### PR TITLE
Make refined type printing more source compatible

### LINKED ISSUE #12939: make refined type printing more source compatible
## Compiler version

3.0.1-RC1

## Minimized example

```Scala
class Bag extends reflect.Selectable

val m = new Bag { val f = 23; def g = 47; def h(i: Int): Int = i }
```

## Output

in the repl the type printed has a type that is not source compatible
```scala
scala> val m = new Bag { val f = 23; def g = 47; def h(i: Int): Int = i; var i = 101; type N = Int }
val m: Bag{f: Int; g: => Int; h(i: Int): Int; i: Int; i_=(x$1: Int): Unit; N = Int} = anon$1@403364e9
```

## Expectation

we annotate each refinement with if it is a `type`, `def` or `val`

```scala
scala> val m = new Bag { val f = 23; def g = 47; def h(i: Int): Int = i; var i = 101; type N = Int }
val m: Bag{val f: Int; def g: Int; def h(i: Int): Int; def i: Int; def i_=(x$1: Int): Unit; type N = Int} = anon$1@403364e9
```

### PR DESCRIPTION
**Pull request by jan-pieter**
Fixes #12939 
Supersedes #13698

### CODE REVIEW

**Comment by SethTisue** [comment]
on the previous PR, @tgodzik wondered if the formatting could be improved. I think it's out of scope for this PR. Maybe Dale's recent pretty-printing PR (not merged yet) will improve the situation

**Comment by SethTisue** [comment]
marking as draft for now since there are some test failures. probably just need to update some expected outputs to match the new format?

**Comment by jan-pieter** [comment]
> probably just need to update some expected outputs to match the new format?
Most of the failures were indeed due to some unexpected output. I've updated the expected output for those.

The _catchall_ case was actually being hit by the tests for capture checking. I lack knowledge about the implementation details to explain that. For now I've updated this case to return an empty string which makes the tests pass. Some example output with the previous implementation:
```
-- [E007] Type Mismatch Error: tests/neg-custom-args/captures/i15772.scala:20:49 ---------------------------------------
20 |    val boxed1 : (({*} C) => Unit) -> Unit = box1(c)  // error
   |                                             ^^^^^^^
   |Found:    {c} 
   |  ({*} 
   |    ({c} 
   |      C{
   |        AnnotatedType(TypeRef(ThisType(TypeRef(NoPrefix,module class <empty>)),class C),CaptureAnnotation(BiMapped3802(Var3741{TermRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),module class caps$)),val *)}, elems = {TermRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),module class caps$)),val *)}),false)) 
   |          arg
   |        : {*} C
   |      }
   |    )
   |   -> Unit)
   | -> Unit
   |Required: (({*} C) => Unit) -> Unit
   |
   | longer explanation available when compiling with `-explain`
```

File: compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala [hunk 1]
```diff
@@ -112,7 +112,13 @@ class PlainPrinter(_ctx: Context) extends Printer {
 
   /** String representation of a refinement */
   protected def toTextRefinement(rt: RefinedType): Closed =
-    (refinementNameString(rt) ~ toTextRHS(rt.refinedInfo)).close
+    val keyword = rt.refinedInfo match {
+      case _: TypeRef => "val "
+      case _: ExprType | _: MethodType => "def "
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala:117]: there could also be PolyType, e.g. `def bar: Foo { def foo[T <: Any](t: T): String } = ???`

Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala:117]: for this case, `Foo` must be a class that already defines a polymorphic method

**Comment by jan-pieter** [comment]
Thanks for your comments and suggestions @bishabosha. I've updated the code & tests accordingly. This also solves the issue with `AnnotatedType`s in the capture checking tests.

**Review by bishabosha — APPROVED** [review]

**Comment by jan-pieter** [comment]
@SethTisue Shall I rebase on the main branch, squash the commits and force push this branch?

**Comment by dwijnand** [comment]
Is it just me that's bothered that `C1{val x: Int}` isn't spaced out as `C1 { val x: Int }`?

**Comment by jan-pieter** [comment]
No, I prefer the latter one as well. However we previously said it to be out of scope for this PR: https://github.com/lampepfl/dotty/pull/16303#issuecomment-1307525828 . This PR only adds keywords. Perhaps https://github.com/lampepfl/dotty/pull/16277 did bring some improvements already? Otherwise I propose to improve that in a separate PR.

**Comment by jan-pieter** [comment]
@SethTisue It took a few tries, but now it's ready to be merged I think :wink:

**Review by dwijnand — APPROVED** [review]

**Comment by SethTisue** [comment]
Yay!

**Pull request merged**